plugin/src/main/scala-3/chisel3/internal/plugin/BundlePhase.scala

```diff
@@ -0,0 +1,227 @@
+        tupleTree(List(tpd.Literal(Constant(name)), sel))
+    }
+    currentFields
+  }
+
+  def generateElements(record: tpd.TypeDef)(using Context): tpd.DefDef = {
+    val bundleSym = record.symbol.asClass
+
+    val currentFields: List[Tree] = getBundleFields(record)
+
+    val dataTpe = requiredClassRef("chisel3.Data")
+    val rhs = makeArray(currentFields)
+
+    val tupleTpe =
+      defn.TupleClass.typeRef.appliedTo(List(defn.StringType, defn.AnyType))
+
+    val iterableTpe =
+      requiredClassRef("scala.collection.Iterable").appliedTo(tupleTpe)
+
+    val elementsSym: Symbol =
+      newSymbol(
+        bundleSym,
+        Names.termName("_elementsImpl"),
+        Flags.Method | Flags.Override | Flags.Protected,
+        MethodType(Nil, Nil, iterableTpe)
+      )
+
+    tpd.DefDef(elementsSym.asTerm, rhs)
+  }
+}
+
+class ChiselBundlePhase extends PluginPhase {
+  val phaseName: String = "chiselBundlePhase"
+  override val runsAfter = Set(TyperPhase.name)
+
+  override def transformTypeDef(record: tpd.TypeDef)(using Context): tpd.Tree = {
+    if (
+      ChiselTypeHelpers.isRecord(record.tpe)
+      && !record.symbol.flags.is(Flags.Abstract)
+    ) {
+      val isBundle: Boolean = ChiselTypeHelpers.isBundle(record.tpe)
+      val thiz:     tpd.This = tpd.This(record.symbol.asClass)
+
+      // // ==================== Generate _cloneTypeImpl ====================
+      val conArgs: Option[List[List[tpd.Tree]]] =
+        BundleHelpers.extractConArgs(record, thiz, isBundle)
+      val cloneTypeImplOpt =
+        BundleHelpers.generateAutoCloneType(record, thiz, conArgs, isBundle)
+
+      // ==================== Generate val elements (Bundles only) ====================
+      val elementsImplOpt: Option[tpd.DefDef] =
+        if (isBundle) Some(BundleHelpers.generateElements(record)) else None
+
+      // ==================== Generate _usingPlugin ====================
+      val usingPluginOpt =
+        if (isBundle) {
+          val isPluginSym = newSymbol(
+            record.symbol,
+            Names.termName("_usingPlugin"),
+            Flags.Method | Flags.Override | Flags.Protected,
+            defn.BooleanType
+          )
+          Some(
+            tpd.DefDef(
+              isPluginSym.asTerm,
+              _ => tpd.Literal(Constant(true))
+            )
+          )
+        } else None
+
+      // TODO
+      // val autoTypenameOpt =
+      //   if (BundleHelpers.isAutoTypenamed(record.symbol)) {
+      //     BundleHelpers.generateAutoTypename(record, thiz, conArgs.map(_.flatten))
+      //   } else None
+
+      record match {
+        case td @ TypeDef(name, tmpl: tpd.Template) => {
+          val newDefs =
+            elementsImplOpt.toList ++ usingPluginOpt.toList ++ cloneTypeImplOpt.toList
+          val newTemplate =
+            if (tmpl.body.size >= 1)
+              cpy.Template(tmpl)(body = newDefs ++ tmpl.body)
+            else
+              cpy.Template(tmpl)(body = newDefs)
```

Comment:
This is a tiny nit, but it's better to prepend to Lists than to append, and you can prepend or append anything iterable (Options included so I'd write newDefs as:
```scala
val newDefs = elementsImplOpt ++: usingPluginOpt ++: cloneTypeImplOpt.toList
```

Similar thing for `tmpl.body`, it's probably better to use `++:`.

That being said, do we really want these new defs at the start of the definition rather than the end?

And do we really need the body size check? Shouldn't it work even if empty?

Reply:
The body size guard I copy-pasted from dotty internals so not exactly sure why it's needed but when I remove it I get a cryptic crash from pickler
```
[error] java.lang.AssertionError: assertion failed: unresolved symbols: value j (line 31) #19714, value k (line 36) #19719, value i (line 29) #19712, value bar in class BundleFooBar (line 11) #9152, value j (line 24) #19320, value i (line 35) #19718, value j (line 37) #19720, value foo in class BundleFooBar (line 10) #9151, value k (line 23) #19319, value i (line 22) #19318, value k (line 30) #19713 when pickling chisel3/src/test/scala/chiselTests/Basic.scala 
```

Don't really wanna dig into this atm so I'll leave it as is for now